tests/MongoDB.Driver.Core.Tests/Core/Helpers/MockClusterableServerFactory.cs

```diff
@@ -74,6 +79,45 @@ public IClusterableServer CreateServer(ClusterId clusterId, IClusterClock cluste
                             {
                                 _servers[mockMonitor.Object.Description.EndPoint].HasBeenRemoved = true;
                             });
+                    Action<string, TopologyDescription?> mockMonitorInvalidate = (reason, responseTopologyVersion) =>
+                    {
+                        var baseDescription = description;
+                        var oldDescription = mockMonitor.Object.Description;
+                        var newDescription = baseDescription.With(
+                            $"InvalidatedBecause:{reason}",
+                             lastUpdateTimestamp: DateTime.UtcNow,
+                             topologyVersion: responseTopologyVersion);
+                        mockMonitor.SetupGet(m => m.Description).Returns(newDescription);
+                        var serverDescriptionChangedEventArgs = new ServerDescriptionChangedEventArgs(
+                            oldDescription,
+                            newDescription);
+                        mockMonitor.Raise(m => m.DescriptionChanged += null, serverDescriptionChangedEventArgs);
+                    };
+                    mockMonitor
+                        .Setup(m => m.Invalidate(It.IsAny<string>(), It.IsAny<TopologyDescription?>()))
+                        .Callback(mockMonitorInvalidate);
+                    var mockConnection = new Mock<IConnectionHandle>();
+                    var mockConnectionPool = new Mock<IConnectionPool>();
+                    var poolGeneration = 0;
+                    var connectionGeneration = 0;
+                    // need to use a func to close over connectionGeneration
+                    mockConnection.Setup(c => c.Generation).Returns(valueFunction: () => connectionGeneration);
+                    // need to use a func to close over poolGeneration
+                    mockConnectionPool.Setup(p => p.Generation).Returns(valueFunction: () => poolGeneration);
```

Comment:
As noted in the comments, without using a func, we fail to close over the value in the "Returns" clause. For more information, see https://justin-pihony.blogspot.com/2013/10/moq-returns-variable-does-not-update.html?m=1